candig/server/ontology.py

```diff
@@ -0,0 +1,219 @@
+        terms = validator.get_duo_list()
+
+    """
+
+    def __init__(self, ont, input_json):
+        self.ontology_file_object = ont
+        self.input_json = input_json
+        self.ids_require_datetime_modifier = ["DUO:0000024"]
+        self.ids_supported = [
+            "DUO:0000001", "DUO:0000002", "DUO:0000003", "DUO:0000004",
+            "DUO:0000005", "DUO:0000006", "DUO:0000007", "DUO:0000011",
+            "DUO:0000012", "DUO:0000014", "DUO:0000015", "DUO:0000016",
+            "DUO:0000017", "DUO:0000018", "DUO:0000019", "DUO:0000020",
+            "DUO:0000021", "DUO:0000022", "DUO:0000024", "DUO:0000025",
+            "DUO:0000026", "DUO:0000027", "DUO:0000028", "DUO:0000029",
+            "DUO:0000042"
+        ]
+        self.ids_need_modifiers_with_def = {
+            "DUO:0000024": "This DUO Term requires you specify date as YYYY-MM-DD format in the modifier attribute."
+        }
+
+    def validate_duo(self):
+        validity = True
+        duo_list = self.input_json.get('duo')
+
+        # if duo_list is empty or None, abort validation right away
+        if not duo_list:
+            return False
+
+        for duo in duo_list:
+            try:
+                duo_id = duo.get("id")
+                modifier = duo.get("modifier")
+
+                # Check if the ID exists, a KeyError will be thrown is it doesn't
+                OntologyParser(self.ontology_file_object, duo_id)
+
+                # Fail is the ID is None
+                if duo_id is None:
+                    validity = False
+                    print("Please specify 'id' for all DUO terms.")
+
+                # Fail if the ID supplied is unsupported
+                if duo_id not in self.ids_supported:
+                    validity = False
+                    print(duo_id, "is not currently supported.")
+
+                # Fail if the modifier is not supplied with IDs that require it
+                if duo_id in self.ids_need_modifiers_with_def and modifier is None:
+                    validity = False
+                    print(duo_id, self.ids_need_modifiers_with_def[duo_id])
+
+                # Fail if the datetime supplied with the IDs that require it is invalid
+                if duo_id in self.ids_require_datetime_modifier:
+                    if self.validate_date_time(modifier) is False:
+                        validity = False
+                        print(duo_id, "has malformed datetime", modifier, "it should be YYYY-MM-DD")
+
+                # Fail if modifier is supplied with IDs that do not require it
+                if modifier is not None and duo_id not in self.ids_need_modifiers_with_def:
+                    validity = False
+                    print(duo_id, "cannot accept a modifier.")
+
+            except KeyError:
+                # Fail if the ID cannot be found in ontology
+                validity = False
+                print("One or more DUO IDs you provide are not valid.")
```

Comment:
We might want to return the duo_id, thus user will know what is the problem.

Reply:
I modified the parser a bit in datasets to do that, but it does change the response slightly. Not sure if you want to use it here, Jimmy.
https://github.com/CanDIG/datasets_service/blob/DUO/candig_dataset_service/ontologies/duo.py

Reply:
i just realized that with the whitelist, this check is unnecessary